File: localizationActivity/src/main/java/com/akexorcist/localizationactivity/core/LocalizationUtility.java
package com.akexorcist.localizationactivity.core;

import android.content.Context;
import android.content.res.Configuration;
import android.os.Build;
import android.os.LocaleList;

import java.util.Locale;

/**
 * Created by Akexorcist on 10/19/2017 AD.
 */

public class LocalizationUtility {
    public static Locale getLocaleFromConfiguration(Configuration configuration) {
        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.N) {
            return configuration.getLocales().get(0);
        } else {
            return configuration.locale;
        }
    }

    public static Context applyLocalizationContext(Context baseContext) {
        Locale baseLocale = LocalizationUtility.getLocaleFromConfiguration(baseContext.getResources().getConfiguration());
        Locale currentLocale = LanguageSetting.getLanguage(baseContext);
        if (!baseLocale.toString().equalsIgnoreCase(currentLocale.toString())) {
            LocalizationContext context = new LocalizationContext(baseContext);
            Configuration config = context.getResources().getConfiguration();
            if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.N) {
                config.setLocale(currentLocale);
                LocaleList localeList = new LocaleList(currentLocale);
                LocaleList.setDefault(localeList);
                config.setLocales(localeList);
                return context.createConfigurationContext(config);
            } else if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.JELLY_BEAN_MR1) {
                config.setLocale(currentLocale);
                return context.createConfigurationContext(config);
            } else {
                return context;
            }
        } else {
            return baseContext;
        }
    }
}

File: localization/src/main/java/com/akexorcist/localizationactivity/core/LocalizationUtility.kt
package com.akexorcist.localizationactivity.core

import android.content.Context
import android.content.res.Configuration
import android.content.res.Resources
import android.os.Build
import android.os.LocaleList
import android.util.DisplayMetrics
import java.util.*

object LocalizationUtility {
    fun getLocalizedContext(baseContext: Context): Context {
        val (configuration, isChanged) = getLocalizedConfiguration(baseContext, baseContext.resources.configuration)
        return when {
            isChanged && Build.VERSION.SDK_INT >= Build.VERSION_CODES.JELLY_BEAN_MR1 ->
                baseContext.createConfigurationContext(configuration)
            isChanged -> {
                @Suppress("DEPRECATION")
                baseContext.resources.updateConfiguration(configuration, baseContext.resources.displayMetrics)
                baseContext
            }
            else -> baseContext
        }
    }

    fun getLocalizedResources(baseContext: Context, baseResources: Resources): Resources {
        val (configuration, isChanged) = getLocalizedConfiguration(baseContext, baseResources.configuration)
        return when {
            isChanged && Build.VERSION.SDK_INT >= Build.VERSION_CODES.JELLY_BEAN_MR1 ->
                baseContext.createConfigurationContext(configuration).resources
            isChanged -> {
                val metrics: DisplayMetrics = baseContext.resources.displayMetrics
                @Suppress("DEPRECATION")
                return Resources(baseContext.assets, metrics, configuration)
            }
            else -> baseResources
        }
    }

    fun getLocalizedConfiguration(baseContext: Context, baseConfiguration: Configuration): Pair<Configuration, Boolean> {
        val defaultLocale = LanguageSetting.getDefaultLanguage(baseContext)
        val currentLocale = LanguageSetting.getLanguageWithDefault(baseContext, defaultLocale)
        val baseLocale = getLocaleFromConfiguration(baseConfiguration)
        if (!isRequestedLocaleChanged(baseLocale, currentLocale)) return baseConfiguration to false
        return when {
            Build.VERSION.SDK_INT >= Build.VERSION_CODES.O -> {
                val localeList = LocaleList(currentLocale)
                LocaleList.setDefault(localeList)
                Configuration(baseConfiguration).apply {
                    setLocale(currentLocale)
                    setLocales(localeList)
                } to true
            }
            Build.VERSION.SDK_INT >= Build.VERSION_CODES.JELLY_BEAN_MR1 -> {
                Configuration(baseConfiguration).apply {
                    setLocale(currentLocale)
                } to true
            }
            else -> {
                Configuration(baseConfiguration).apply {
                    @Suppress("DEPRECATION")
                    this.locale = currentLocale
                    if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.JELLY_BEAN_MR1) {
                        setLayoutDirection(currentLocale)
                    }
                } to true
            }
        }
    }

    @Suppress("DEPRECATION")
    private fun getLocaleFromConfiguration(configuration: Configuration): Locale {
        return if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.O) {
            configuration.locales.get(0) ?: Locale.getDefault()
        } else {
            configuration.locale
        }
    }

    private fun isRequestedLocaleChanged(baseLocale: Locale, currentLocale: Locale) =
        !baseLocale.toString().equals(currentLocale.toString(), ignoreCase = true)
}
